### PR TITLE
Fix commit history.

### PR DESCRIPTION
**Pull request by graham-jackson**
# Description of change
Add support for incremental updates.
https://themindgym.atlassian.net/browse/DB-170

# QA steps
 - [y ] automated tests passing
 
# Risks

# Rollback steps
 - revert this branch

### CODE REVIEW

File: tap_mavenlink/streams/base.py [hunk 1]
```diff
@@ -13,14 +13,21 @@
...
     KEY_PROPERTIES = ['id']
     CACHE = False
 
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        self.stream_metadata = get_stream_metadata(self.catalog.to_dict())
+        self.replication_method = self.stream_metadata.get('replication-method')
+        self.replication_key = self.stream_metadata.get('replication-key')
+
```
Comment (graham-jackson) [inline tap_mavenlink/streams/base.py:30]: @adrien-mg comment from first (bad) attempt at this PR
> Using this syntax I think it's possible that some streams end up with: self.replication_method == None and self.replication_key == None if those fields are not defined explicitly in the catalog's stream's metadata?
> Not sure if this is a problem at all though.

You're right. This shouldn't be a problem though, since the args validation at the top of the app will only allow sensible method-key combinations. That's the plan anyway 🤞

File: tap_mavenlink/streams/base.py [hunk 2]
```diff
@@ -61,11 +79,20 @@ def sync_paginated(self, params, extra):
...
 
+            self.save_state(transformed[-1])
+
             if page_number >= total_pages:
                 break
             else:
                 params['page'] += 1
 
+    def save_state(self, last_record):
+        # TODO is this overkill since we know we don't have compisite keys here?
```
Comment (graham-jackson) [inline tap_mavenlink/streams/base.py]: > What would be overkill in this comment? The check whether the replication_key is in the last_record?

Good point! This TODO is no longer relevant (since we explicitly prevent composite replication-keys as part of arg validation). 

I'll remove the TODO.

File: tap_mavenlink/utils.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+            if bookmarks[bookmark]['field'] != BOOKMARK_FIELD:
+                raise ValueError(f'The only bookmark field implemented is: {BOOKMARK_FIELD}')
+
+            try:
+                date_text = bookmarks[bookmark]['last_record']
+                datetime.datetime.strptime(date_text, DATETIME_PARSE)
+            except ValueError:
+                raise ValueError("Unable to parse as datetime.")
+
+    if args.catalog:
```
Comment (graham-jackson) [inline tap_mavenlink/utils.py]: > I think this check is superfluous as there should always be a catalog given as a --catalog flag when calling a tap.

Good point, I'll remove this.

File: tap_mavenlink/streams/base.py (hunk 1 above)
Comment (adrien-mg) [reply tap_mavenlink/streams/base.py:30]: :+1:

File: tap_mavenlink/streams/base.py (hunk 2 above)
Comment (adrien-mg) [reply tap_mavenlink/streams/base.py]: :+1:

File: tap_mavenlink/utils.py (hunk 3 above)
Comment (adrien-mg) [reply tap_mavenlink/utils.py]: :+1:

**Comment by adrien-mg** [comment]
LGTM. Very well done PR and apologies for the delay.
* Pytest usage is spot-on
* Tests setup readable and gives a good dev experience
* Can't think of any corner cases we're missing

Can you please bump the version to `0.0.4` in the `setup.py` file before merging? 

Good work man :pray:

**Review by adrien-mg — APPROVED** [review]

**Review by adrien-mg — APPROVED** [review]
